explorer/chartscache.go

```diff
@@ -0,0 +1,403 @@
+package explorer
+
+import (
+	"context"
+	"encoding/gob"
+	"os"
+	"sync"
+
+	"github.com/decred/dcrdata/db/dbtypes"
+	"github.com/decred/dcrdata/txhelpers"
+)
+
+// Cache data for charts that appear on /charts page is managed here.
+
+// chartDataCache is a data cache for the historical charts that appear on
+// /charts page on the UI.
+type chartDataCache struct {
+	sync.RWMutex
+	updateHeight int64
+	Data         map[string]*dbtypes.ChartsData
+}
+
+// cacheChartsData holds the prepopulated data that is used to draw the charts.
+var cacheChartsData chartDataCache
+
+// get provides a thread-safe way to access the all cache charts data.
+func (c *chartDataCache) get() map[string]*dbtypes.ChartsData {
+	c.RLock()
+	defer c.RUnlock()
+	return c.Data
```

Comment:
The map may still be accessed concurrently.  e.g. `prePopulateChartsData` reads from the map while `DropToAncestor` reads and writes.  `get()` just controls access to the `Data` field of `chartDataCache`, not the map itself.